File: src/io/DumpXYZ.py
#  Copyright (C) 2016
#      Max Planck Institute for Polymer Research & JGU Mainz
#  Copyright (C) 2012,2013
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.


r"""
*********************
espressopp.io.DumpXYZ
*********************

* `dump()`

  write configuration to trajectory XYZ file. By default filename is ``out.xyz``,
  coordinates are folded. DumpXYZ works also for Multiple communicators.

  **Properties**

* `filename`
  Name of trajectory file. By default trajectory file name is ``out.xyz``

* `unfolded`
  False if coordinates are folded, True if unfolded. By default - False

* `append`
  True if new trajectory data is appended to existing trajectory file. By default - True

* `length_factor`
  If length dimension in current system is nm, and unit is 0.23 nm, for example, then
  ``length_factor`` should be 0.23
  Default: 1.0

* `length_unit`
  It is length unit. Can be ``LJ``, ``nm`` or ``A``. By default - ``LJ``

* `store_pids`
    True if you want to store pids as fastwritexyz does. False otherwise (standard XYZ)
    Default: False

* `store_velocities`
    True if you want to store velocities. False otherwise (XYZ doesn't require it)
    Default: False

usage:

writing down trajectory

>>> dump_conf_xyz = espressopp.io.DumpXYZ(system, integrator, filename='trajectory.xyz')
>>> for i in range (200):
>>>   integrator.run(10)
>>>   dump_conf_xyz.dump()

writing down trajectory using ExtAnalyze extension

>>> dump_conf_xyz = espressopp.io.DumpXYZ(system, integrator, filename='trajectory.xyz')
>>> ext_analyze = espressopp.integrator.ExtAnalyze(dump_conf_xyz, 10)
>>> integrator.addExtension(ext_analyze)
>>> integrator.run(2000)

Both examples will give the same result: 200 configurations in trajectory .xyz file.

setting up length scale

For example, the Lennard-Jones model for liquid argon with :math:`\sigma=0.34 [nm]`

>>> dump_conf_xyz = espressopp.io.DumpXYZ(system, integrator, filename='trj.xyz', \
>>>                                       unfolded=False, length_factor=0.34, \
>>>                                       length_unit='nm', store_pids=True, \
>>>                                       store_velocities = True, append=True)

will produce trj.xyz with in nanometers

.. function:: espressopp.io.DumpXYZ(system, integrator, filename=out.xyz, unfolded=False,\
                                    length_factor=1.0, length_unit='LJ', store_pids=False,\
                                    store_velocities=False, append=True)

	:param system:
	:param integrator:
	:param filename:
	:param bool unfolded:
	:param real length_factor:
	:param length_unit:
	:param bool store_pids:
	:param bool store_velocities:
	:param bool append:
	:type system:
	:type integrator:
	:type filename:
	:type length_unit:

.. function:: espressopp.io.DumpXYZ.dump()

		:rtype:
        
"""

from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.ParticleAccess import *
from _espressopp import io_DumpXYZ

class DumpXYZLocal(ParticleAccessLocal, io_DumpXYZ):

  def __init__(self, system, integrator, filename='out.xyz', unfolded=False, length_factor=1.0, length_unit='LJ', store_pids=False, store_velocities=False, append=True):
    cxxinit(self, io_DumpXYZ, system, integrator, filename, unfolded, length_factor, length_unit, store_pids, store_velocities, append)

  def dump(self):
    if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
      self.cxxclass.dump(self)


if pmi.isController :
  class DumpXYZ(ParticleAccess):
    __metaclass__ = pmi.Proxy
    pmiproxydefs = dict(
      cls =  'espressopp.io.DumpXYZLocal',
      pmicall = [ 'dump' ],
      pmiproperty = ['filename', 'unfolded', 'length_factor', 'length_unit', 'store_pids', 'store_velocities', 'append']
    )

File: src/io/DumpXYZ.cpp
/*
  Copyright (C) 2016
      Max Planck Institute for Polymer Research & JGU Mainz
  Copyright (C) 2012-2015
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI

  This file is part of ESPResSo++.

  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "python.hpp"
#include <fstream>
#include <iomanip>
#include <iostream>
#include "analysis/ConfigurationExt.hpp"
#include "analysis/ConfigurationsExt.hpp"
#include "DumpXYZ.hpp"
#include "storage/Storage.hpp"
#include "bc/BC.hpp"

using namespace espressopp;
using namespace espressopp::analysis;
using namespace std;

namespace espressopp {
  namespace io {

    void DumpXYZ::dump(){

      shared_ptr<System> system = getSystem();
      ConfigurationsExt conf( system );
      conf.setUnfolded(unfolded);
      conf.gather();

      if( system->comm->rank()==0 ){
        ConfigurationExtPtr conf_real = conf.back();

        int num_of_particles = conf_real->getSize();

        char *ch_f_name = new char[file_name.length() + 1];
        strcpy(ch_f_name, file_name.c_str());
        ofstream myfile (ch_f_name, ios::out | ios::app);
        if (myfile.is_open()){
          myfile << num_of_particles << endl;

          Real3D Li = system->bc->getBoxL();

          // for noncubic simulation boxes
          myfile << Li[0] * length_factor << "  0.0  0.0  0.0  "<<
                  Li[1] * length_factor << "  0.0  0.0  0.0  "<< Li[2] * length_factor;
          // additional info to comment line
          myfile << "  currentStep " << integrator->getStep() << "  lengthUnit "<< length_unit << endl;

          ConfigurationExtIterator cei = conf_real-> getIterator();
          std::streamsize p = myfile.precision();
          for (size_t i=0; i<num_of_particles; i++) {

                if (store_pids)  {
                    myfile << cei.currentId() << " ";
                }

                myfile << particleIDToType.find(cei.currentId())->second <<
                       " " << fixed << setprecision(10) << length_factor * cei.currentProperties()[0] << " "
                       << length_factor * cei.currentProperties()[1] <<
                       " " << length_factor * cei.currentProperties()[2];

                if (store_velocities) {
                  myfile << " " << length_factor * cei.currentProperties()[3] << " " << length_factor * cei.currentProperties()[4] <<
                  " " << length_factor * cei.currentProperties()[5];
                }
                myfile << endl;
                myfile.unsetf(ios_base::fixed);
                myfile << setprecision(p);

                cei.incrementIterator();
          }

          myfile.close();
        }
        else cout << "Unable to open file: "<< file_name <<endl;

        delete [] ch_f_name;
      }
    }

    // Python wrapping
    void DumpXYZ::registerPython() {

      using namespace espressopp::python;

      class_<DumpXYZ, bases<ParticleAccess>, boost::noncopyable >
      ("io_DumpXYZ", init< shared_ptr< System >,
                           shared_ptr< integrator::MDIntegrator >,
                           std::string,
                           bool,
                           real,
                           std::string ,
                           bool,
                           bool,
                           bool>())
        .add_property("filename", &DumpXYZ::getFilename,
                                  &DumpXYZ::setFilename)
        .add_property("unfolded", &DumpXYZ::getUnfolded,
                                  &DumpXYZ::setUnfolded)
        .add_property("length_factor", &DumpXYZ::getLengthFactor,
                                       &DumpXYZ::setLengthFactor)
        .add_property("length_unit", &DumpXYZ::getLengthUnit,
                                     &DumpXYZ::setLengthUnit)
        .add_property("store_pids", &DumpXYZ::getStorePids,
                                     &DumpXYZ::setStorePids)
        .add_property("store_velocities", &DumpXYZ::getStoreVelocities,
                                     &DumpXYZ::setStoreVelocities)
        .add_property("append", &DumpXYZ::getAppend,
                                  &DumpXYZ::setAppend)
        .def("dump", &DumpXYZ::dump)
      ;
    }
  }
}
